### PR TITLE
Add `/oauth/inspect` endpoint

### PR DESCRIPTION
**Pull request by jeanp413**
## Description

Add `/oauth/inspect` endpoint

## Related Issue(s)

Related https://github.com/gitpod-io/gitpod/issues/12254

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

File: components/server/src/oauth-server/db.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ function createVSCodeClient(protocol: "vscode" | "vscode-insiders" | "vscodium")
             { name: "function:getWorkspace" },
             { name: "function:getWorkspaces" },
             { name: "function:getSSHPublicKeys" },
-            { name: "function:sendHeartBeat" },
+            // { name: "function:sendHeartBeat" },
```
Comment (jeanp413) [inline components/server/src/oauth-server/db.ts:76]: For testing

File: components/server/src/oauth-server/oauth-controller.ts [hunk 2]
```diff
@@ -175,6 +176,17 @@ export class OAuthController {
             }
         });
 
+        router.get("/oauth/inspect", async (req: express.Request, res: express.Response) => {
+            const clientId = req.query.client as string;
+            if (typeof clientId !== "string" || !Object.keys(inMemoryDatabase.clients).includes(clientId)) {
+                res.sendStatus(400);
```
Comment (akosyakov) [inline components/server/src/oauth-server/oauth-controller.ts:182]: return?

File: components/server/src/oauth-server/oauth-controller.ts [hunk 3]
```diff
@@ -175,6 +176,17 @@ export class OAuthController {
             }
         });
 
+        router.get("/oauth/inspect", async (req: express.Request, res: express.Response) => {
+            const clientId = req.query.client as string;
+            if (typeof clientId !== "string" || !Object.keys(inMemoryDatabase.clients).includes(clientId)) {
```
Comment (AlexTugarev) [inline components/server/src/oauth-server/oauth-controller.ts:181]: this definitely has the potential to provide an unguarded endpoint to scan for random `clientId`s.

tagging as security relevant.

Comment (akosyakov) [reply components/server/src/oauth-server/oauth-controller.ts:181]: everything is already available in this repo? 🤔

Comment (jeanp413) [reply components/server/src/oauth-server/oauth-controller.ts:181]: yeah client ids are already public

**Comment by akosyakov** [comment]
@jeanp413 Let's open it for review?

**Review by mustard-mh — APPROVED** [review]
/hold 

hold to squash commits

Code LGTM, tested in browser

File: components/server/src/oauth-server/oauth-controller.ts (hunk 3 above)
Comment (easyCZ) [reply components/server/src/oauth-server/oauth-controller.ts:181]: I think the point Alex is making is that this allows you to enumerate all client IDs we may have. Therefore, you can estimate how big of an installation of gitpod this is, how many users it has and which exact IDs exist in the system. 

There's a difference between the client ID being public and knowing exactly how many client IDs there are.

Comment (akosyakov) [reply components/server/src/oauth-server/oauth-controller.ts:181]: What would be an alternative? cc @geropl

Comment (akosyakov) [reply components/server/src/oauth-server/oauth-controller.ts:181]: > Therefore, you can estimate how big of an installation of gitpod this is, how many users it has and which exact IDs exist in the system.

I'm not sure how you can do it though. You can only there Gitpod installation supports installing ssh keys or not. You cannot get any other information from this endpoint.

Comment (easyCZ) [reply components/server/src/oauth-server/oauth-controller.ts:181]: Would you be able to provide a use-case this is looking to solve? Is it to discover my own ClientID?

Comment (akosyakov) [reply components/server/src/oauth-server/oauth-controller.ts:181]: Please have a look at this [RFC](https://www.notion.so/gitpod/Gitpod-Server-API-compatibility-check-ee00fcfe60a84c1e8e018ab6c06c1f5d#5aed369a61cf46c7bf945ae017a6e125) It is a work around for how we do auth now.

Comment (geropl) [reply components/server/src/oauth-server/oauth-controller.ts:181]: @mikenikles  @AlexTugarev  I think there is a misunderstanding here: Clients !== API clients.

Client is a possible _kind_ of clients we support as defined [here](https://github.com/gitpod-io/gitpod/blob/79ef436f154158d4a4c78aa7e619cedfd52ee83d/components/server/src/oauth-server/db.ts#L38-L96).

> everything is already available in this repo?

:point_up: Enumerating those is fine I think. :slightly_smiling_face:

**Review by geropl — APPROVED** [review]
LGTM. Thx @jeanp413 ! :+1:

**Pull request merged**